### DISCUSSION TITLE
Issues with running migrations programatically

### DISCUSSION THREAD

**Discussion by revmischa** [prisma/prisma#8245] [Q&A]
I'm trying to figure out how to run prisma migrations using AWS Lambda. I am not using the Serverless framework, I am using [this template](https://github.com/jetbridge/typescript-cdk-template) ([why](https://spiegelmock.com/2021/05/29/frameworkless-web-applications-aws-cdk/)) and trying to adapt it to Prisma.

I don't really understand how I can run migrations. My idea is to create a lambda function that runs the migrations programatically. I don't really see an API for doing so. I tried to cram the CLI migration script into a [NodejsFunction](https://docs.aws.amazon.com/cdk/api/latest/docs/aws-lambda-nodejs-readme.html) CDK construct but I have encountered a problem doing that. When I run the script I get:
```json
{
  "errorType": "Runtime.ImportModuleError",
  "errorMessage": "Error: Cannot find module '../package.json'\nRequire stack:\n- /var/task/index.js\n- /var/runtime/UserFunction.js\n- /var/runtime/index.js",
  "trace": [
    "Runtime.ImportModuleError: Error: Cannot find module '../package.json'",
    "Require stack:",
    "- /var/task/index.js",
    "- /var/runtime/UserFunction.js",
    "- /var/runtime/index.js",
    "    at _loadUserApp (/var/runtime/UserFunction.js:100:13)",
    "    at Object.module.exports.load (/var/runtime/UserFunction.js:140:17)",
    "    at Object.<anonymous> (/var/runtime/index.js:43:30)",
    "    at Module._compile (internal/modules/cjs/loader.js:1068:30)",
    "    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1097:10)",
    "    at Module.load (internal/modules/cjs/loader.js:933:32)",
    "    at Function.Module._load (internal/modules/cjs/loader.js:768:16)",
    "    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:72:12)",
    "    at internal/main/run_main_module.js:17:47"
  ]
}
```

Looking through `@prisma/migrate` I see a number of places where it tries to require `../package.json`. This does not work with JavaScript bundling because all dependencies are bundled into a single file, there is no `node_modules` directory structure. AWS's official `NodeJsFunction` uses [esbuild](https://esbuild.github.io/) to bundle everything into an `index.js`.

## Questions:
* How should I be running DB migrations in a serverless (not [Serverless](https://www.serverless.com/)) environment?
* Can I run migrations programmatically without needing a shell?
* Can `@prisma/migrate` not explode on import if it's part of a bundle and doesn't have a `../package.json` file in the local path?

## Migration script:
This is what I hacked together:
```ts
import { Migrate } from "@prisma/migrate"
import { ensureDatabaseExists } from "@prisma/migrate/dist/utils/ensureDatabaseExists"
import { printFilesFromMigrationIds } from "@prisma/migrate/dist/utils/printFiles"
// import { getSchemaPath } from "@prisma/sdk"
import chalk from "chalk"

export const handler = async (): Promise<string> => {
  const schemaPath = "./prisma/schema.prisma"
  console.log("path", schemaPath)

  const migrate = new Migrate()
  const wasDbCreated = await ensureDatabaseExists("apply", true, schemaPath)
  if (wasDbCreated) {
    console.info() // empty line
    console.info(wasDbCreated)
  }

  const diagnoseResult = await migrate.diagnoseMigrationHistory({
    optInToShadowDatabase: false,
  })
  const listMigrationDirectoriesResult = await migrate.listMigrationDirectories()

  if (listMigrationDirectoriesResult.migrations.length > 0) {
    const migrations = listMigrationDirectoriesResult.migrations
    console.info(`${migrations.length} migration${migrations.length > 1 ? "s" : ""} found in prisma/migrations`)
  } else {
    console.info(`No migration found in prisma/migrations`)
  }

  const editedMigrationNames = diagnoseResult.editedMigrationNames
  if (editedMigrationNames.length > 0) {
    console.info(
      `${chalk.yellow("WARNING The following migrations have been modified since they were applied:")}
${editedMigrationNames.join("\n")}`
    )
  }

  const { appliedMigrationNames: migrationIds } = await migrate.applyMigrations()

  migrate.stop()

  console.info() // empty line
  if (migrationIds.length === 0) {
    return chalk.greenBright(`No pending migrations to apply.`)
  } else {
    return `The following migration${migrationIds.length > 1 ? "s" : ""} have been applied:\n\n${chalk(
      printFilesFromMigrationIds("migrations", migrationIds, {
        "migration.sql": "",
      })
    )}
      
${chalk.greenBright("All migrations have been successfully applied.")}`
  }
}
```

And the CDK construct:
```ts
export class MigrationScript extends Construct {
  constructor(scope: Construct, id: string, { vpc, db }: DatabaseScriptProps) {
    super(scope, id)

    // migrate script
    const func = new NodejsFunction(this, "migrate", {
      timeout: Duration.seconds(20), // todo: config
      memorySize: 2048,
      projectRoot: `${__dirname}/..`,
      environment: {
        NODE_OPTIONS: "--enable-source-maps",
      },
      runtime: Runtime.NODEJS_14_X,
      bundling: {
        sourceMap: true,
        commandHooks: {
          beforeInstall: () => [],
          afterBundling: () => [],
          beforeBundling: (inputDir: string, outputDir: string): string[] => {
            return [
              // need to copy over migration files
              `cp -r "${__dirname}/../../../../repo/prisma" "${outputDir}"`,
              `cp -r "${__dirname}/../package.json" "${outputDir}"`,
            ]
          },
        },
      },
      vpc,
    })

    // allow DB access
    db.connections.allowDefaultPortFrom(func)
  }
}
```

**Comment by ryands17** [comment]
Currently we [do not support running migrate programmatically yet](https://github.com/prisma/prisma/issues/4703). You can use `exec` to run `prisma migrate` and that would give you the same output.

As for bundling, I would suggest keeping `prisma` and `@prisma/client` both in `dependencies` in your `package.json` so that they are bundled and you can add this to `bundling` to make sure that they are not packaged by esbuild and stay as `node_modules`.

```ts
bundling: {
  sourceMap: true,
  nodeModules: ['@prisma/client', 'prisma']
}
```

Then running `prisma migrate` via `exec` should work.

**Reply by revmischa** [reply]
Thanks that was helpful information!
What I did is create a lambda layer - https://github.com/jetbridge/lambda-layer-prisma-pg

However now my problem is the size of prisma - specifically `@prisma/engines`

<img width="676" alt="Screen Shot 2021-07-13 at 6 06 29 PM" src="https://user-images.githubusercontent.com/245131/125476385-85b68efb-a372-4ab9-8600-c1124867ce98.png">

So now I can't even deploy my function
`Resource handler returned message: "Function code combined with layers exceeds the maximum allowed size of 262144000 bytes. The actual size is 266859163 bytes.`

What's really weird is my function bundle is only 44kb and the layer is 140MB unzipped, so I'm not sure where the 266MB is coming from but that's probably not prisma's fault.

**Reply by ryands17** [reply]
I would suggest removing all the engines except for the `migration-engine` as that's just what is required in this case. This can be done in the `afterBundling` hook with a `rm -r`.

**Reply by revmischa** [reply]
I'm trying to deploy my normal (non-migration-script) lambdas with CDK. I want to deploy the minimum of the generated client with my function code and not the other deps which I can get from the layer. I'm not clear how to include `@prisma/client` without bringing in `@prisma/engines`.

I don't want to use a bundle hook because I want this to be a generic solution.

Also I'm not sure I see my models in the generated bundle anywhere

My `externalModules` is 
```ts
[
  'prisma',
  '@prisma/migrate',
  '@prisma/engines',
  'pg',
  'pg-native',
  'aws-sdk'
]
```

And I have `nodeModules: ["@prisma/client"]`.
Alas my function is too big:

<img width="697" alt="Screen Shot 2021-07-14 at 12 55 59 PM" src="https://user-images.githubusercontent.com/245131/125602892-934c256d-13ac-43be-b3b6-fb23c1877bc3.png">


`Resource handler returned message: "Unzipped size must be smaller than 114950977 bytes`

I tried `nodeModules: [".prisma/client"],`
But get:
`Error: Cannot extract version for module '.prisma/client'. Check that it's referenced in your package.json or installed.`

**Reply by janpio** [reply]
Prisma Client does not need `@prisma/engines` at all as there is a copy of the needed Engine (Query Engine)  in the generated Client at `.prisma/client` in `node_modules`. If you keep thins separate, one function for Migrations and one for Client - you can completely separate these things and have 2 sets of dependencies. (`@prisma/engines` is only required for Prisma CLI, which you probably have to run the Migrations.

**Reply by revmischa** [reply]
I don't need to bundle `@prisma/engines` at all in my function because it's provided by [the layer](https://github.com/jetbridge/lambda-layer-prisma-pg). But the combination of `nodeModules: ["@prisma/client"]` and everything else in `externalModules` (see above) still bundles `node_modules/@prisma/engines` with my function which makes it very large.

My goal here is to find a lean way to bundle a function that has my prisma client with https://docs.aws.amazon.com/cdk/api/latest/docs/aws-lambda-nodejs-readme.html . I assume that I should build a layer and then only include the prisma generated client with my function bundle, but I haven't discovered the right combination of arguments yet to do this.

**Reply by ryands17** [reply]
There's no other option I can think of other than removing it after bundling as I have done [here](https://github.com/ryands17/prisma-lambda/blob/main/lib/prisma-lambda-stack.ts#L35-L36). `@prisma/engines` will be installed regardless so it's best to remove it in a bundling hook.

**Reply by revmischa** [reply]
> There's no other option I can think of other than removing it after bundling as I have done [here](https://github.com/ryands17/prisma-lambda/blob/main/lib/prisma-lambda-stack.ts#L35-L36). `@prisma/engines` will be installed regardless so it's best to remove it in a bundling hook.

Thanks for your help @ryands17 @janpio - I am making progress with it.
I was partially successful using:
```ts
bundling: {
    minify: true,
    // sourceMap: true,
    // metafile: true,
    nodeModules: ["@prisma/client", "prisma"], //["@prisma/client"],
    externalModules: [
      // from layer
      "prisma",
      "@prisma/migrate",
      "@prisma/engines",
      "pg",
      "pg-native",

      // from lambda env
      "aws-sdk",
    ],
    commandHooks: {
      beforeBundling: (_inputDir: string, _outputDir: string) => [],
      beforeInstall: (_inputDir: string, outputDir: string) => [`cp -R ${__dirname}/../../repo/prisma ${outputDir}/`],
      afterBundling: (_inputDir: string, outputDir: string) => {
        return [
          `cd ${outputDir}`,
          `npx prisma generate`,
          `rm -rf node_modules/@prisma/engines`,
          `rm -rf node_modules/@prisma/client/node_modules node_modules/.bin node_modules/prisma`,
        ]
      },
    },
  }
```

I've included my [Prisma layer](https://github.com/jetbridge/lambda-layer-prisma-pg). The layer is public and contains [these packages](https://github.com/jetbridge/lambda-layer-prisma-pg/blob/main/layer/package.json). The layer works when I use it to run migrations.

Layer contents:
<img width="668" alt="Screen Shot 2021-07-14 at 4 09 19 PM" src="https://user-images.githubusercontent.com/245131/125627471-7e264cd3-bf43-4d80-ba83-0a84d6b8317b.png">

But when I run a function I get this error:
```ts
        "Error: ",
        "Invalid `prisma.tenant.findMany()` invocation:",
        "",
        "",
        "  Query engine binary for current platform \"rhel-openssl-1.0.x\" could not be found.",
        "This probably happens, because you built Prisma Client on a different platform.",
        "(Prisma Client looked in \"/var/task/node_modules/@prisma/client/runtime/query-engine-rhel-openssl-1.0.x\")",
        "",
        "Searched Locations:",
        "",
        "  /var/task/node_modules/.prisma/client",
        "  /Users/cyber/dev/jb/freezy/freezy-backend/packages/infra/cdk.out/bundling-temp-db4d5d20e33d01bbceab0c5c62d479b1cfc6ec1ffc0ed9f07ba5bfb7330ff020/node_modules/@prisma/client",
        "  /var/task/node_modules/@prisma/client",
        "  /var/task/node_modules/.prisma/client",
        "  /var/task/prisma",
        "  /tmp/prisma-engines",
        "  /var/task/node_modules/.prisma/client",
        "",
        "",
        "To solve this problem, add the platform \"rhel-openssl-1.0.x\" to the \"binaryTargets\" attribute in the \"generator\" block in the \"schema.prisma\" file:",
        "generator client {",
        "  provider      = \"prisma-client-js\"",
        "  binaryTargets = [\"native\"]",
        "}",
        "",
        "Then run \"prisma generate\" for your changes to take effect.",
        "Read more about deploying Prisma Client: https://pris.ly/d/client-generator",
        "    at cb (/var/task/node_modules/@prisma/client/runtime/index.js:34806:17)",
        "    at get (/var/task/index.js:604:295113)",
        "    at Runtime.Stn.dispatch [as handler] (/var/task/index.js:466:12220)"
```

Which seems odd as the layer appears to contain it and the layer is attached to my function.

<img width="1502" alt="Screen Shot 2021-07-14 at 4 10 34 PM" src="https://user-images.githubusercontent.com/245131/125627628-e72ec278-2eb0-4992-89f5-292381d4ff90.png">

**Reply by revmischa** [reply]
My function bundle looks like:

<img width="708" alt="Screen Shot 2021-07-14 at 4 11 37 PM" src="https://user-images.githubusercontent.com/245131/125627778-1ff2b833-9388-4f17-8163-d1ccb4efeb95.png">

**Reply by janpio** [reply]
You have the "wrong" engines, or at least in the wrong location. For running the Client, you just need `@prisma/client` and  `.prisma/client` (with the generated Client, which includes the correct query engine binary in the correct location (usually `/var/task/node_modules/.prisma/client`). The layer is providing the engines for the CLI (so Migration things for example).

Usually our Prisma Client can be completely isolated and standalone without the Prisma CLI, Prisma Engines or any other packages after it is generated. It can be packaged up as a Npm package and then used in the project without any further dependencies of Prisma.

**Reply by revmischa** [reply]
Thanks, I figured out this problem, sort of, I think.
My functions need the engines - the migration script needs the migration engine and the normal functions need the query engine.

The migration script imports and runs utilities in `@prisma/migrate` which is fully contained in the layer. Thus the migration script is able to find the engine because its search path is in `/opt/nodejs/node_modules/...` which is the layer path, which contains the engines.

My normal functions were unable to find the query engine because they were using a different search path, `/var/task/node_modules/`, which is the function bundle.

I solved this by setting `PRISMA_QUERY_ENGINE_BINARY: "/opt/nodejs/node_modules/@prisma/engines/query-engine-rhel-openssl-1.0.x"` which works but is another hack I had to discover to get this thing to run.

I'd be happy to use the prisma client standalone, I don't mind using the layer if that's needed (although 140MB for some binaries seems a bit steep). 

I am trying to push my work in progress to https://github.com/jetbridge/jetkit-cdk/pull/13 - I am hoping to put together all my discoveries into a reusable library and also to show what I had to do to get things running in CDK. I am optimistic that there is a simpler approach than what I'm doing.
Feedback on that PR would be very much welcome. I was hoping seeing it altogether would make it clearer what I'm trying to accomplish and share my efforts so that others in the future can be spared having to work through these bundling issues themselves.

**Reply by janpio** [reply]
Hm right, you connected the Prisma Client part with the Prisma Migrate part doing this - which kinda works.

When you do `npx prisma generate` you should be getting a client that is self contained. Usually people do that in their off serverless build pipeline, and then upload the resulting client with their function code to Lambda. (A very manual example how this works: https://github.com/prisma/e2e-tests/tree/dev/platforms-serverless/lambda) Am I missing something about your project and how you are doing things that prevents you from doing just that? (If you really want to use a layer, you can probably also put the generated Client only into another layer that is used for the "access the database" functions you have - vs. the "migrate the database" functions that use the layer which has all the engines) 

Only for the Migrate part the layer _should_ become relevant. (You could remove/delete all the non migration engines for that layer btw if you really just use it for the "migrate the database" functions and make it a lot smaller that way)

**Reply by revmischa** [reply]
> When you do npx prisma generate you should be getting a client that is self contained. 

This sounds nice, right now I'm using the official CDK NodejsFunction bundle which builds it on my mac resulting in the wrong binary. It is possible to force it to run in docker but that will slow things down and I should be able to use the engine from my layer right? I plan on having many functions so keeping the bundle size and time to build down is very desireable.

For my non-migration functions, these settings worked for me:
```ts
// default Lambda settings
const functionOptions: FunctionOptions = {
  tracing: Tracing.ACTIVE, // x-ray
  environment: {
    PRISMA_QUERY_ENGINE_BINARY: "/opt/nodejs/node_modules/@prisma/engines/query-engine-rhel-openssl-1.0.x",
  },
  bundling: {
    minify: true,
    sourceMap: true,
    externalModules: ["pg", "pg-native", "aws-sdk"],
    nodeModules: ["@prisma/client", "prisma"], //["@prisma/client"],
    externalModules: [
      // from layer
      "prisma",
      "@prisma/migrate",
      "@prisma/engines",
      "pg",
      "pg-native",

      // from lambda env
      "aws-sdk",
    ],
    commandHooks: {
      beforeBundling: (_inputDir: string, _outputDir: string) => [],
      beforeInstall: (_inputDir: string, outputDir: string) => [`cp -R ${__dirname}/../../repo/prisma ${outputDir}/`],
      afterBundling: (_inputDir: string, outputDir: string) => {
        return [
          `cd ${outputDir}`,
          `npx prisma generate`,
          `rm -rf node_modules/@prisma/engines`,
          `rm -rf node_modules/.prisma/client/*-engine-*`,
          `rm -rf node_modules/@prisma/client/node_modules node_modules/.bin node_modules/prisma`,
        ]
      },
    },
  },
}
```

**Reply by janpio** [reply]
If you are getting the wrong engines as you are building on a different OS, you can use `binaryTargets` in your schema. The error message _should_ tell you about that :hmmm:. Here are the docs: https://www.prisma.io/docs/concepts/components/prisma-schema/generators#binary-targets-prisma-client-js

**Reply by revmischa** [reply]
I did see that but I assumed bundling the engine would increase my function size and I plan to have dozens of functions so would prefer to simply use the layer. Is my assumption correct?

**Reply by janpio** [reply]
Kinda. If you have many functions, it might make sense to put the full generated Client into a layer and use that across these functions. But splitting generated Client and the used Engine apart introduces a high risk of incompatibility and so on that will be hard to impossible to debug. I would really try to keep it the code and Engine together on one level - either function, or layer, but not cross boundaries.

**Reply by revmischa** [reply]
Okay thanks I take your point. I wonder how hard it would be to make a CDK construct that generates this layer automatically as part of the stack? I think it'd require the right combination of tsconfig/package.json/bundling props but should be possible

**Reply by janpio** [reply]
That I do not know much about honestly. Layers and CDK are quite advanced, so everyone seems to have very specific requirements that make it hard to generalize the solutions. If you get to one, please make sure to share it. Having a simple way to put your generated Client into a layer that is used across functions sounds like an awesome improvement in that regard.

**Reply by revmischa** [reply]
Definitely, I'm doing my best to generalize and share my solutions in https://github.com/jetbridge/jetkit-cdk/pull/13

**Comment by revmischa** [comment]
One unfortunate side effect of using `npx prisma generate` in the bundle hook is that it runs for every function I'm deploying, and I plan to have a lot of functions. 
In a perfect world it would generate it once and bundle only that generated node module with my function with everything else living in the layer.
I'm working on adding this all to https://github.com/jetbridge/freezy-backend/pull/21

**Comment by janpio** [comment]
Quick question parallel to our "figuring technical things out" comment thread above:

_Why_ do you want to run Migrations via a serverless functions? Usually our users do that via the Prisma CLI in a build step that happens on CI or on their local machine. Do you have a use case where this would not work?

**Reply by revmischa** [reply]
> _Why_ do you want to run Migrations via a serverless functions? Usually our users do that via the Prisma CLI in a build step that happens on CI or on their local machine. Do you have a use case where this would not work?

I need to run the migrations against my Aurora Serverless database. Since there is no data API support (would make my life a lot easier) I need to run them somewhere with network access to the database in my VPC. Is there a better way to achieve this? I use GitHub actions, I'm not clear how I would run migrations from there or my local machine.

**Reply by janpio** [reply]
Network access via VPC is a very valid answer here. 

I do not think we have a better answer than to use a bastion host inside the VPC, connect to that one via SSH and then run the commands from there. (Might still be less complexity that running Migrations via serverless directly)

**Reply by revmischa** [reply]
I've got it set up now in JetKit so that all you have to do in your stack is:
```ts
import { DatabaseMigrationScript } from "@jetkit/cdk"

new DatabaseMigrationScript(this, "MigrationScript", {
        vpc,
        db,
        prismaPath: `${__dirname}/../../repo/prisma`,
      })
```
And it constructs a lambda that runs migrations for you and returns the output as a string. You can invoke the lambda however you want.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
